### ISSUE TITLE
Error message “! Argument of \xpg@nospace has an extra }.” with polyglossia

### ISSUE THREAD

**Issue by Alfonsib** [reutenauer/polyglossia#52]
I use a code – to be found at the end –  for long citations in the french tradition. When I use french quotation marks inside such a long citation, I have no problem on compiling with xetex if I load babel with the french option, and the error message “! Argument of \xpg@nospace has an extra }.” when loading polyglossia. The same happens on compiling with lualatex, though with a different error message, beginning with 
“! Package soul Error: Reconstruction failed.”

The problem disappears if I define in the preamble of the document :
\newcommand_{\og}{«{}} \newcommand_{\fg}{{}»}
and if the french quotation marks inside the long citation are replaced by \og … \fg.

So I suspect there is some conflict between polyglossia and soul – or my way of using soul. I first posted on fr.comp.text.tex, but obtained no solution.

Best regards,
B. Alfonsi

Code (and minimal example):

\documentclass[a4paper,11pt, twoside]{article} 

\usepackage{fontspec}% f 
\setmainfont{Minion Pro}% 
%\usepackage[french]{babel} 
\usepackage{polyglossia} 
\setmainlanguage{french} 

\usepackage{soulutf8} 
\usepackage {kvsetkeys} 
\makeatletter 
\newsavebox \citeguill@box 
\define@key {citeguill}{left}[\guillemotleft]{\def\citeguill@left {#1}} 
\define@key {citeguill}{right}[\guillemotright]{\def\citeguill@right {#1}} 
\define@key {citeguill}{middle}[\hskip -\fontdimen2\font\mbox \guillemotright\hskip\fontdimen2\font]{\def\citeguill@middle 
 {\setbox\citeguill@box =\hbox {#1}\def\citeguill@middle {\copy 
\citeguill@box }}% 
}% \KV@citeguill@middle 
\newcommand*\citeguillsetup [1]{\def\citeguill@setup {\kvsetkeys 
{citeguill}{#1}}} 
\let\citeguill@setup =\relax % <init> 

\newcommand*\citeguill [1][]{% 
 \kvsetkeys{citeguill}{left,right,middle}\citeguill@setup% 
\kvsetkeys{citeguill}{#1}% 
 \SOUL@setup% 
 \def\SOUL@preamble {\citeguill@middle \citeguill@left }% 
 \def\SOUL@postamble {\citeguill@right }% 
 \def\SOUL@everysyllable {\the\SOUL@syllable }% 
 \def\SOUL@everyspace ##1{##1\nobreak \ 
 \discretionary {\kern-\fontdimen 2\font } 
 {\citeguill@middle } 
 {}}% 
 \def\SOUL@everyhyphen {\discretionary% 
 {\SOUL@setkern \SOUL@hyphkern \SOUL@sethyphenchar } 
 {\citeguill@middle } 
 {}}% 
 \def\SOUL@everyexhyphen ##1{\SOUL@setkern \SOUL@hyphkern 
 \hbox{##1}\discretionary {} 
 {\citeguill@middle } 
 {\SOUL@setkern \SOUL@charkern }}% 
 \SOUL@% 
}% \citeguill 
\makeatother 

\begin{document} 

  Pour conclure, les rédacteurs de \emph{L’Anti-Mythes} résument bien la question BHL telle qu’elle se pose à l’automne 1977 : \citeguill{Que le cynisme séduise, que la bêtise paie et que la mode soit au renouveau des idées les plus réactionnaires, cela ne suffit pas à expliquer l’extraordinaire exploitation par les médias du phénomène «nouveaux philosophes», ni leur impressionnant succès commercial.} 

\end{document}

**Comment by jspitz** [comment]
xelatex compiles fine with current head. lualatex still fails with the mentioned error (“! Package soul Error: Reconstruction failed.”)

**Comment by jspitz** [comment]
For the record, it's the  `\xpg@frpt=1\relax` in gloss-french triggering this.

**Comment by wehro** [comment]
There is a slight difference between the spacings introduced by xelatex and lualatex: Xelatex inserts a `\thinspace`, lualatex a `\kern`. Soul is not able to handle kerns, as you can see by trying something like `\ul{a\kern1ptb}`.

**Comment by jspitz** [comment]
`polyglossia-fprt.lua` inserts the \kern, right? Can we make it insert proper space, instead? This would also solve the problem that it does not catch redefinitions of `\thinspace` and friends.

**Comment by wehro** [comment]
> `polyglossia-fprt.lua` inserts the `\kern`, right?

Yes, this file inserts a “kern node” (in terms of LuaTeX) with a width of 0.16667 em, where XeTeX inserts a `\thinspace`, which is defined as `\def\thinspace{\leavevmode@ifvmode\kern .16667em }` in `latex.ltx`.

> Can we make it insert proper space, instead? This would also solve the problem that it does not catch redefinitions of `\thinspace` and friends.

I fear, this is not possible without substantially changing the code of `polyglossia-frpt.lua`. So far, this file serves as `pre_linebreak_filter` callback of LuaTeX (cf. page 172 of the LuaTeX manual). This means, it comes into operation just before TeX breaks the paragraph into lines. All macros have to be expanded at this time and it is only possible to insert certain basic node types, but no arbitrary TeX code.

**Comment by jspitz** [comment]
If I am reading frenchb.lua (of babel-french) correctly, this adds a 0x202F (NARROW NO-BREAK SPACE) instead. Would this be an option?

**Comment by wehro** [comment]
I think so, but I have not yet tried it.

**Comment by jspitz** [comment]
@dohyunkim do you happen to have an idea how to fix this?

**Comment by wehro** [comment]
Please do not change anything in `polyglossia-frpt.lua` until the new Latin language module will be merged. Improvements will be welcome thereafter.

**Comment by jspitz** [comment]
I suppose @dohyunkim is aware of that, see #302

**Comment by dohyunkim** [comment]
I just attempted the babel-french way.
```diff
diff --git a/tex/polyglossia-frpt.lua b/tex/polyglossia-frpt.lua
index 9d0ae23..c15ce43 100644
--- a/tex/polyglossia-frpt.lua
+++ b/tex/polyglossia-frpt.lua
@@ -136,6 +136,16 @@ local function set_spacings(thinsp, thicksp)
   thinspace = thinsp
 end
 
+local function get_white_glyph(g, wd)
+  local nbsp = node_copy(g)
+  if wd == thinspace then
+    nbsp.char = 0x202F -- NARROW NO-BREAK SPACE
+  else
+    nbsp.char = 0x00A0 -- NO-BREAK SPACE
+  end
+  return nbsp
+end
+
 -- from typo-spa.lua
 local function process(head)
     local done = false
@@ -166,8 +176,12 @@ local function process(head)
                                 head = remove_node(head,prev)
                             end
                         end
+                        if attr == 1 then -- default xpg@frpt attribute
+                          insert_node_before(head, start, get_white_glyph(start, map[2]))
+                        else
                           insert_node_before(head,start,get_penalty_node())
                           insert_node_before(head,start,get_kern_node(map[2]*quad))
+                        end
                         done = true
                     end
                     local next = start.next
@@ -186,8 +200,12 @@ local function process(head)
                                 head = remove_node(head,next)
                             end
                         end
+                        if attr == 1 then -- default xpg@frpt attribute
+                          insert_node_after(head, start, get_white_glyph(start, map[2]))
+                        else
                           insert_node_after(head,start,get_kern_node(map[2]*quad))
                           insert_node_after(head,start,get_penalty_node())
+                        end
                         done = true
                     end
                 end
```

With lualatex, no error occured now and the result seems to be ok. Only we get messages like this in the log: 
```
Missing character: There is no   (U+202F) in font [lmroman10-regular]:+tlig;!
``` 

With luahblatex-dev and using `mode=harf` (or `Renderer=OpenType`) font option, we get no missing-character messages. Harfbuzz seems to do the work to be done.

As shown in the patch, the original method, inserting kern node, will be activated when `\xpg@frpt=2` is declared.

**Comment by jspitz** [comment]
Excellent. Let's play with that after @wehro's stuff has settled.

**Comment by dohyunkim** [comment]
While digging into the issue, I suddenly realized that the implementation for luatex and that for xetex are different from each other. 

* Currently polyglossia-frpt ignores `\hbox` constructions, only using `pre_linebreak_filter` callback. 
* For xetex, gloss-french inserts whitespace before the first punctuation in a box, and the same after the last punctuation. But polyglossia-frpt ignores them.

So here is another patch, which does not use NBSP glyphs.
```diff
diff --git a/tex/polyglossia-frpt.lua b/tex/polyglossia-frpt.lua
index 9d0ae23..409a31a 100644
--- a/tex/polyglossia-frpt.lua
+++ b/tex/polyglossia-frpt.lua
@@ -153,7 +153,8 @@ local function process(head)
                 if map then
                     local quad = get_quad(start.font) -- might be optimized
                     local prev = start.prev
-                    if map[1] == left and prev then
+                    if map[1] == left then
+                      if prev then
                         local prevprev = prev.prev
                         local somespace = somespace(prev,true)
                         -- TODO: there is a question here: do we override a preceding space or not?...
@@ -169,9 +170,14 @@ local function process(head)
                         insert_node_before(head,start,get_penalty_node())
                         insert_node_before(head,start,get_kern_node(map[2]*quad))
                         done = true
+                      else
+                        head = insert_node_before(head,start,get_kern_node(map[2]*quad))
+                        done=true
+                      end
                     end
                     local next = start.next
-                    if map[1] == right and next then
+                    if map[1] == right then
+                      if next then
                         local nextnext = next.next
                         local somepenalty = somepenalty(next,10000)
                         if somepenalty then
@@ -189,6 +195,10 @@ local function process(head)
                         insert_node_after(head,start,get_kern_node(map[2]*quad))
                         insert_node_after(head,start,get_penalty_node())
                         done = true
+                      else
+                        insert_node_after(head,start,get_kern_node(map[2]*quad))
+                        done = true
+                      end
                     end
                 end
             end
@@ -203,19 +213,23 @@ local function process(head)
     return head, done
 end
 
-local callback_name = "pre_linebreak_filter"
+local callback_names = { "pre_linebreak_filter", "hpack_filter" }
 
 local function activate()
+  for _, callback_name in ipairs(callback_names) do
     if not priority_in_callback (callback_name, "polyglossia-frpt.process") then
       add_to_callback(callback_name, process, "polyglossia-frpt.process", 1)
     end
   end
+end
 
 local function desactivate()
+  for _, callback_name in ipairs(callback_names) do
     if priority_in_callback (callback_name, "polyglossia-frpt.process") then
       remove_from_callback(callback_name, "polyglossia-frpt.process")
     end
   end
+end
 
 polyglossia.activate_frpt    = activate
 polyglossia.desactivate_frpt = desactivate
```

**Comment by wehro** [comment]
>     Currently polyglossia-frpt ignores `\hbox` constructions, only using `pre_linebreak_filter` callback.

This is a bug I also realized some time ago.
It makes no sense to discuss the code now, as I have introduced some changes. Please let's wait.

**Comment by jspitz** [comment]
The merger's done, so we can go ahead with this.

**Comment by jspitz** [comment]
Fixed by 7e1d20dff1385
Thanks, @dohyunkim!
